pennylane/optimize/rotoselect.py

```diff
@@ -0,0 +1,189 @@
+
+    Keyword Args:
+        possible_generators (list[~.Operation]): List containing the possible ``pennylane.ops.qubit`` operators
+            that are allowed in the circuit. Default is the set of Pauli matrices :math:`\{X, Y, Z\}`.
+
+    **Example:**
+
+    Initialize the Rotoselect optimizer, set the initial values of  the weights ``x``,
+    choose the initial generators, and set the number of steps to optimize over.
+
+    >>> opt = qml.optimize.RotoselectOptimizer()
+    >>> x = [0.3, 0.7]
+    >>> generators = [qml.RX, qml.RY]
+    >>> n_steps = 1000
+
+    Set up the PennyLane circuit using the ``default.qubit`` as simulator device.
+
+    >>> dev = qml.device("default.qubit", analytic=True, wires=2)
+
+    >>> @qml.qnode(dev)
+    >>> def circuit(params, generators=None):  # generators will be passed as a keyword arg
+    >>>     generators[0](params[0], wires=0)
+    >>>     generators[1](params[1], wires=1)
+    >>>     qml.CNOT(wires=[0, 1])
+    >>>     return qml.expval(qml.PauliX(0)), qml.expval(qml.PauliY(1))
+
+    Define a cost function based on the above circuit.
+
+    >>> def cost(x, generators):
+    >>>     X_1, Y_2 = circuit(x, generators=generators)
+    >>>     return 0.2 * X_1 + 0.5 * Y_2
+
+    Run the optimization step-by-step for ``n_steps`` steps.
+
+    >>> cost_rotosel = []
+    >>> for _ in range(n_steps):
+    >>>     cost_rotosel.append(cost(x, generators))
+    >>>     x, generators = opt.step(cost, x, generators)
+
+    The optimized values for x should now be stored in ``x`` together with the optimal gates for
+    the circuit, while steps-vs-cost can be seen by plotting ``cost_rotosel``.
+    """
+    # pylint: disable=too-few-public-methods
+
+    def __init__(self, possible_generators=None):
+        self.possible_generators = possible_generators or [qml.RX, qml.RY, qml.RZ]
+
+    def step(self, objective_fn, x, generators):
+        r"""Update x with one step of the optimizer.
+
+        Args:
+            objective_fn (function): The objective function for optimization. It must have the
+                signature ``objective_fn(x, generators=None)`` with a sequence of the values ``x``
+                and a list of the gates ``generators`` as inputs, returning a single value.
+            x (Union[Sequence[float], float]): Sequence containing the initial values of the
+                variables to be optimized over, or a single float with the initial value.
+            generators (list[~.Operation]): List containing the initial ``pennylane.ops.qubit``
+                operators to be used in the circuit and optimized over.
+
+        Returns:
+            array: The new variable values :math:`x^{(t+1)}` as well as the new generators.
+        """
+        x_flat = np.fromiter(_flatten(x), dtype=float)
+
+        for d, _ in enumerate(x_flat):
+            x_flat[d], generators[d] = self._find_optimal_generators(
+                objective_fn, x_flat, generators, d
+            )
```

Comment:
I'm guessing that `x_flat[d]` is the parameter of gate `generators[d]`? In which case it might be worth raising an exception if `len(generators) != len(x_flat)`.

Reply:
That's a good point. 👍